erpnext/manufacturing/doctype/bom/bom.py

```diff
@@ -113,7 +113,7 @@ def get_bom_material_detail(self, args=None):
 		item = self.get_item_det(args['item_code'])
 		self.validate_rm_item(item)
 
-		args['bom_no'] = args['bom_no'] or item and cstr(item[0]['default_bom']) or ''
+		args['bom_no'] = item and cstr(item[0]['default_bom']) or ''
```

Comment:
but what if the item is same, but the user wants a non-default BOM?

this must be handled at the UI. When item is changed, clear the BOM and then reset it.